### ISSUE TITLE
Incorrect digits in zeta zeros

### ISSUE THREAD

**Issue by jwj61** [LMFDB/lmfdb#1293]
From the issue tracker:

"Last digits of every imaginary part of Riemann zeta zeros are incorrect. For example, first zero ends with ....24766, but according to http://oeis.org/A058303, it should end with ...247027. Usually 2-4 last digits are incorrect throughout the whole list. In http://lmfdb.org/zeros/zeta/HowComputed it is said that the zeros are accurate to +-2^-102 ~ 30 digits after the decimal point, so I would suggest to mention this on the main site (http://lmfdb.org/zeros/zeta/), otherwise the list is misleading.

Best wishes,
Marcin"

Please respond to mszynisz@gmail.com

**Comment by davidfarmer** [comment]
When we fix them, we should truncate (not round), and also exlpain
that we are doing that.

For the web page version can we just truncate the strings
as the table is populated?

**Comment by JohnCremona** [comment]
The disclaimer text has been deployed, but the issue stays open until the incorrect digits are not displayed. 
@arbooker

**Comment by arbooker** [comment]
This issue has long been known and there is no perfect resolution. The zeros are stored in binary but displayed in decimal, so there will always be issues with the tablemaker's dilemma, unless we are willing to truncate many places occasionally, which is a needless loss of precision.

My favo(u)red solution is to display round(t*10^30)/10^30. That guarantees that the error is at most 1 in the last place. @jwbober

**Comment by arbooker** [comment]
Another option is to display round(t*10^31)/10^31. Then the approximation is off by no more than 3 in the last place, and it has the added feature that you could recover the exact internal representation from the decimal approximation. I think Odlyzko does something similar in his zeta zero tables.

**Comment by jwbober** [comment]
The zeros should be printed so that when they are corrected rounded to 104 bits of absolute precision, the internal representation is recovered, which will be accurate to +/- 2^(-102). (Perhaps there are some digits in the internal representation that are always zero. I don't know. I'm sure that Dave chose 104 because it is a multiple of 8.) Due to double rounding, this probably means that the displayed decimal number is not necessarily accurate to 2^(-101).

In any case, I'm printing too many digits. An offending line (there is more than one) is at https://github.com/LMFDB/lmfdb/blob/master/lmfdb/zeros/zeta/platt_zeros.py#L78
```python
    mpmath.mp.prec = log(t1, 2) + 10 + 101  # We make sure that the working precision
                                            # is large enough. Note that we are adding
                                            # a little too much here, so when these
                                            # numbers are printed, they will have too
                                            # many digits.
```

Andy's suggestions seem good. In this case I probably prefer to round to 31 digits, if that makes the original number recoverable. In other cases I decided to display however many correct digits we have in base 10, except for some special cases (like a zero that is 0, except that we don't know that it is exactly 0). That would probably work pretty well here, but we would be discarding some information, and, as Andy notes, we would discard a lot of information for zeros that happen to be very close to an integer divided by a power of ten.

**Comment by arbooker** [comment]
I suggest working with the rational numbers directly rather than as floating point. Then the functions
t --> round(t*10^31)/10^31
and
t --> round(t*2^101)/2^101
are inverse to each other when applied to Dave's zeros.

Also, it would make sense to keep trailing zeros, so every number has exactly 31 digits following the decimal point. This makes it clearer what we mean when we say that the error is at most 3 in the last place.

**Comment by jwbober** [comment]
Yes, I agree. I will get to that eventually, but maybe not today.

I was lazy when I wrote this before, and didn't want to think much about
how to print.

**Comment by davidfarmer** [comment]
I think some of the suggestions miss the important point that there
are two distinct uses for the zeros:

1) Visually looking at the number in a browser, and maybe cutting and
pasting a small number of then to doing something silly.

2) Doing something serious with them.

For 2), the user should download them, and it is important that we
not lose any precision.  For this use case, extra meaningless digits
can be there and will cause no harm.

For 1), the audience is completely different.  Such a user is likely
to think in terms if digits, and they may not clearly understand the
description of the precision and why we are showing "wrong" digits.
And if we have 25 instead of 30 digits, it would be all the same to them.

The "View" and "download" links are clearly visible, so nothing needs
to be done for 2).  For 1) we need to change the numbers shown in the
table on the web page.  The following would be perfectly satisfactory
to the typical 1) person, and would be mathematically correct:

14.13472514173469379045725198356...
21.02203963877155499262847959389...
25.01085758014568876321379099256...
etc

So, for the html only, truncate so all digits are correct, and then
stick on dots.  There is a perfect solution, but it involves two separate
imperfect solutions, each of which is satisfactory to a different
audience.

**Comment by arbooker** [comment]
@davidfarmer I understand what you're getting at, but I think it's fair to ask for some level of intelligence on the part of the users of the site. If we say that the numbers are off by at most 3 in the last place, that should be enough of an explanation. This is the one place on the site (so far) where we can give guaranteed accuracy, and we shouldn't throw that away just to be visually appealing. I also disagree that "if we have 25 instead of 30 digits, it would be all the same to them"; there are over 10^11 zeros, which means we can expect to lose 11 places of accuracy at least once if we follow your proposal. Wouldn't you wonder why there was a zero with only 19 digits displayed when most of them have 30?

**Comment by JohnCremona** [comment]
I'm with David on this one but I know nothing about zeros.

**Comment by jwj61** [comment]
I see David's point about there being two use cases.

For people who are browsing on the web page, I don't think we can count on them reading the fine print.  For downloading, definitely give the full precision you have.  For browsing, I don't think you need to worry about making full precision recoverable.  If truncation to number of known decimal digits means that some zeros loose some digits, it is not that big a loss.  In fact, a very small percentage of the stored zeros will be looked at closely via browsing.

**Comment by AndrewVSutherland** [comment]
Is anyone working on this critical issue?  To who should it be assigned?

**Comment by davidfarmer** [comment]
I removed the "critical" designation because of the disclaimer added in pull request #1296 .

**Comment by jwbober** [comment]
I assigned myself to this (now that someone else removed the "critical" tag) and I'll get to it eventually. But of course @davidfarmer is welcome to fix it if he likes.

**Comment by davidfarmer** [comment]
I would do it, but for the added overhead of having to also
install another type of database.

Cremona offered to let me do it in an account on a Warwick
machine, so maybe I should do that.

**Comment by JohnCremona** [comment]
I'll contact @davidfarmer to arracnge this

**Comment by AndrewVSutherland** [comment]
Fixed by #2858
